File: swift/Sources/AlloyAudio/DSP/Effects/Delay.swift
import Foundation

/// Tempo-syncable stereo / ping-pong delay send unit with damped feedback.
/// 100% wet; fed by the delay send tap. Twin of web
/// src/dsp/effects/delay.ts (canonical).

public final class Delay: SendEffect {
    private let params: DelayParams
    private var bufL: [Double]
    private var bufR: [Double]
    private let size: Int
    private var pos = 0
    private let delaySamples: Int
    private var lpL = 0.0
    private var lpR = 0.0
    private let fb: Double
    private let dampCoef: Double
    private let pingpong: Bool

    public init(params: DelayParams, sampleRate: Double) {
        self.params = params
        self.delaySamples = max(1, Int(((params.timeMs / 1000) * sampleRate).rounded()))
        self.size = delaySamples + 1
        self.bufL = [Double](repeating: 0, count: size)
        self.bufR = [Double](repeating: 0, count: size)
        self.fb = params.feedback
        self.dampCoef = params.damping
        self.pingpong = params.mode == .pingpong
    }

    public func reset() {
        for i in 0..<bufL.count { bufL[i] = 0 }
        for i in 0..<bufR.count { bufR[i] = 0 }
        pos = 0
        lpL = 0
        lpR = 0
    }

    public func process(inL: inout [Float], inR: inout [Float], outL: inout [Float], outR: inout [Float], frames: Int) {
        for n in 0..<frames {
            var rp = pos - delaySamples
            if rp < 0 { rp += size }
            let dl = bufL[rp]
            let dr = bufR[rp]

            // Damped feedback (one-pole LPF on the delayed signal).
            lpL += dampCoef * (dl - lpL)
            lpR += dampCoef * (dr - lpR)

            // Feedback routing: ping-pong crosses channels.
            let fbL = pingpong ? lpR : lpL
            let fbR = pingpong ? lpL : lpR

            var wl = Double(inL[n]) + fb * fbL
            var wr = Double(inR[n]) + fb * fbR
            if abs(wl) < 1e-20 { wl = 0 }
            if abs(wr) < 1e-20 { wr = 0 }
            bufL[pos] = wl
            bufR[pos] = wr

            pos += 1
            if pos >= size { pos = 0 }

            // 100% wet output = the delayed taps.
            outL[n] = Float(dl)
            outR[n] = Float(dr)
        }
    }
}

File: web/packages/alloy-audio/src/dsp/effects/delay.ts
// Tempo-syncable stereo / ping-pong delay send unit with damped feedback.
// 100% wet; fed by the delay send tap. Twin: Delay.swift.

import { type DelayParams, type SendEffect } from './effect-types.js';

export class Delay implements SendEffect {
  private readonly bufL: Float32Array;
  private readonly bufR: Float32Array;
  private readonly size: number;
  private pos = 0;
  private readonly delaySamples: number;
  private lpL = 0;
  private lpR = 0;
  private readonly fb: number;
  private readonly dampCoef: number;
  private readonly pingpong: boolean;

  constructor(
    private readonly params: DelayParams,
    sampleRate: number,
  ) {
    this.delaySamples = Math.max(1, Math.round((params.timeMs / 1000) * sampleRate));
    this.size = this.delaySamples + 1;
    this.bufL = new Float32Array(this.size);
    this.bufR = new Float32Array(this.size);
    this.fb = params.feedback;
    this.dampCoef = params.damping;
    this.pingpong = params.mode === 'pingpong';
  }

  reset(): void {
    this.bufL.fill(0);
    this.bufR.fill(0);
    this.pos = 0;
    this.lpL = 0;
    this.lpR = 0;
  }

  process(inL: Float32Array, inR: Float32Array, outL: Float32Array, outR: Float32Array, frames: number): void {
    for (let n = 0; n < frames; n++) {
      let rp = this.pos - this.delaySamples;
      if (rp < 0) rp += this.size;
      const dl = this.bufL[rp];
      const dr = this.bufR[rp];

      // Damped feedback (one-pole LPF on the delayed signal).
      this.lpL += this.dampCoef * (dl - this.lpL);
      this.lpR += this.dampCoef * (dr - this.lpR);

      // Feedback routing: ping-pong crosses channels.
      const fbL = this.pingpong ? this.lpR : this.lpL;
      const fbR = this.pingpong ? this.lpL : this.lpR;

      let wl = inL[n] + this.fb * fbL;
      let wr = inR[n] + this.fb * fbR;
      if (Math.abs(wl) < 1e-20) wl = 0;
      if (Math.abs(wr) < 1e-20) wr = 0;
      this.bufL[this.pos] = wl;
      this.bufR[this.pos] = wr;

      this.pos++;
      if (this.pos >= this.size) this.pos = 0;

      // 100% wet output = the delayed taps.
      outL[n] = dl;
      outR[n] = dr;
    }
  }
}
